submit-api/src/Cardano/TxSubmit/Web.hs

```diff
@@ -25,85 +44,143 @@ import Data.Proxy
+        Left err -> handleDeserialiseErr err
+        Right byronOrShelleyTx -> do
+          resp <- liftIO $ submitTx connectInfo byronOrShelleyTx
+          handleSubmitResult resp
   where
     errorResponse :: ToJSON e => e -> Handler a
     errorResponse e = throwError $ err400 { errBody = Aeson.encode e }
 
-decodeByronTx :: ByteString -> Either DecoderError (GenTx ByronBlock)
-decodeByronTx bs =
-    toGenTx <$> fromCborTxAux bs
-  where
-    toGenTx :: Ledger.ATxAux ByteString -> GenTx ByronBlock
-    toGenTx tx = Byron.ByronTx (Byron.byronIdTx tx) tx
+    -- Log relevant information and return an appropriate HTTP 400 response to
+    -- the client.
+    handleDeserialiseErr :: DecoderError -> Handler a
+    handleDeserialiseErr err = do
+      let serr = if
+                  | BS.length txBytes == 0 -> TxSubmitEmpty
+                  | BS.all isHexOrWhitespace txBytes -> TxSubmitDecodeHex
+                  | otherwise -> TxSubmitDecodeFail err
+      liftIO $ logInfo trce $
+        "txSubmitPost: failed to deserialise transaction: "
+          <> renderTxSubmitWebApiError serr
+      errorResponse serr
 
---TODO: remove this local definition of this function as soon as the updated
--- ledger lib with Byron.fromCborTxAux is available
-fromCborTxAux :: ByteString ->  Either DecoderError (Ledger.ATxAux ByteString)
-fromCborTxAux bs =
-    fmap (annotationBytes lbs)
-      $ Binary.decodeFullDecoder "Cardano.Chain.UTxO.TxAux.fromCborTxAux"
-                                 Binary.fromCBOR lbs
-  where
-    lbs = LBS.fromStrict bs
+    -- Log relevant information, update the metrics, and return the result to
+    -- the client.
+    handleSubmitResult :: Either TxSubmitError TxId -> Handler TxId
+    handleSubmitResult res =
+      case res of
+        Left err -> do
+          liftIO $ logInfo trce $
+            "txSubmitPost: failed to submit transaction: "
+              <> renderTxSubmitError err
+          errorResponse (TxSubmitFail err)
+        Right txid -> do
+          liftIO $ logInfo trce $
+            "txSubmitPost: successfully submitted transaction "
+              <> renderMediumTxId txid
+          liftIO $ Gauge.inc (tsmCount metrics)
+          pure txid
 
-    annotationBytes :: Functor f => LBS.ByteString -> f Binary.ByteSpan -> f ByteString
-    annotationBytes bytes = fmap (LBS.toStrict . Binary.slice bytes)
+    -- Attempt to deserialise either a Byron or Shelley transaction.
+    -- Note that, if this fails, the 'DecoderError' from the Shelley
+    -- transaction will be returned and the one from the Byron transaction
+    -- will be discarded.
+    tryDeserialiseByronOrShelleyTx
+      :: ByteString
+      -> Either DecoderError (Either (Tx Byron) (Tx Shelley))
+    tryDeserialiseByronOrShelleyTx bs =
+      case deserialiseFromCBOR AsByronTx bs of
+        Left _ ->
+          case deserialiseFromCBOR AsShelleyTx bs of
+            Left err -> Left err
+            Right shelleyTx -> Right $ Right shelleyTx
```

Comment:
I suspect the following should work: (though haven't tested it)
```suggestion
          Right <$> deserialiseFromCBOR AsShelleyTx bs
```

Reply:
Right, good catch.